s390x support

iii-i opened the pull request:
Hi,

Please consider pulling s390x architecture support (originally developed as part of moonjit).
This branch was created as follows:

- Applied `luajit-s390x.patch` from `luajit-2.1.0-0.18beta3.fc33.src.rpm` to `v2.1.0-beta3` tag
- Rebased to `v2.1` branch
- Added two fixes

The result passes most of the moonjit testsuite on x86_64 and s390x.
What fails are tests for new language features and nonportable artithmetic, which should be OK.

I've decided to go this way instead of taking commits from the moonjit repo, because Fedora version of LuaJIT can be considered battle-tested, and the goal here is to integrate this stable code upstream with minimum changes possible. The rest of s390x support from moonjit (if any) can go in separately later.

Sarah-Kriesch commented:
Is there any problem with merging this PR, where we can support?

sindrom91 commented:
This doesn't really look finished. It seems to have some files that should not really be there (DynASM tests, bash script), some stuff that is copied from ARM and left unchanged. It is mentioned that moonjit tests are passing, but the question is if [LuaJIT tests](https://github.com/LuaJIT/LuaJIT-test-cleanup) are passing?

I would suggest to clean up the DynASM port first and try to get it in and then make a separate PR for the interpreter. It will probably be easier for maintainer to review that way, since there is a lot of code involved.

**Note that I'm not a maintainer.**

levivic commented:
Can anyone move forward this PR? We have potential products/customers depending on Luajit on ibm s390x platforms.

teward commented:
The Lua module for nginx now requires luajit to function. As a result, it is now impossible to build nginx for s390x arch in Debian because there is zero support for s390x in LuaJIT. Can this get put in a higher priority list for the LuaJIT team?

lysliu commented:
> 

Assume https://github.com/openresty/luajit2 has s390x support, I have s390x native environment, testing with it.
And I will create a PR if needed.

iii-i commented:
I've rebased the code on top of the latest v2.1; squashed all commits; added a short description, which I will quote below for convenience; removed unnecessary debug code and fixed a few stylistic issue. I did not split out DynASM support yet, but I think it's a good idea and I will look into it.

```
Add s390x architecture support

s390x (IBM Z) is an architecture of server computers produced by IBM.
It is supported by a number of open source code generators, such as
GCC, LLVM, OpenJDK, eBPF, QEMU, Valgrind and Cranelift. One of the
missing pieces in the ecosystem support is LuaJIT.

The s390x support for LuaJIT was initially developed by @ketank-new,
@mundaym and @niravthakkar. It found its way into moonjit and luajit2
forks, as well as Fedora distro (as a patch). There were also smaller
contributions by @preetikhorjuvenkar, @Bisht13, @velemas and @iii-i.

This is a cumulative patch of the work mentioned above. It contains
all the contributions squashed together, plus minor stylistic
cleanups. It passes all the tests from LuaJIT-test-cleanup, except
for contents.lua, which fails on x86_64 as well.
```

juru1234 commented:
Tested with neovim on s390x. Works for me.

dev-japo commented:
I've tested it on s390 and on x64.
Note, there is an additional warning on x64 systems:
```
lj_ccall.c: In function ‘ccall_set_args’:
lj_ccall.c:1025:34: warning: variable ‘onstack’ set but not used [-Wunused-but-set-variable]
 1025 |     MSize n, isfp = 0, isva = 0, onstack = 0;
      |                                  ^~~~~~~
```
the `LJ_TARGET_S390X` switch hides the logic not the declaration.

Link against noevim on s390x works as expected no issues found.

Thank you very much @iii-i

mcepl commented:
> Tested with neovim on s390x. Works for me.

Yes, builds on s390x with openSUSE/Factory.

juru1234 commented:
@MikePall Is there a chance that this will be merged soon?

iii-i commented:
* Fixed the warning.
* Split out the DynASM changes.

skriesch commented:
@MikePall Can you merge it, please? It is already tested by different Linux distributions and other open source community members.

NeilHanlon commented:
I have tested this builds and runs well on s390x on RHEL/CentOS Stream

lysliu commented:
I have tested this builds and runs well on s390x on Ubuntu 20.04

dev-japo commented:
Hello @MikePall,
is there a missing requirement, to merge this pull request?

mephi42 commented:
* Rebased.

iii-i commented:
* Rebased.

niklas88 commented:
Dear @MikePall ,

I'm writing to you not in my capacity as an IBM employee but on a Sunday afternoon on my personal time and
as a fellow Open Source developer with personal interest in this PR.  Because of the lack of s390x support in
LuaJIT, I and several of my colleagues have to keep compiling [Neovim](https://github.com/neovim/neovim) from source to use it on s390x.
 That's not a big burden of course but it is annoying considering all the pieces to make things work properly are there.

I believe there has been some conflict between you and IBM in the past but respectfully, this is getting ridiculous.
For 4 years this PR has been continuously updated and rebased. The code has been tested and even added as downstream patches in several distributions with Fedora 41 development trees just as the latest example.
And yet you have not commented on any of the s390x PRs with a single request to change something, a
single objective argument why you can't support the code or anything at all. No one is asking you to like IBM,
or to put unreasonable extra work into this. but I do ask that you give this PR a fair chance especially considering
how very clearly people are willing to step up to help fix any issues it might cause down the road.

In my opinion this is also a question of respect, yes @iii-i is getting paid for this work but he, the people
who worked on earlier versions of this and everyone involved here are real people, putting in real effort, time and
energy. They have been showing a tremendous amount of patience despite the lack of progress. Please Mike treat
that with the respect it deserves not for IBM's sake but out of solidarity with your fellow developers and users.

Thank you,
Niklas Schnelle

mcepl commented:
> For 4 years this PR has been continuously updated and
rebased. The code has been tested and even added as downstream
patches in several distributions with Fedora 41 development trees
just as the latest example.

We (I am a SUSE employee) have this patch in our enterprise products as well.

edsiper commented:
@mcepl is there a SUSE fork with these patches in?

mcepl commented:
> @mcepl is there a SUSE fork with these patches in?

https://build.opensuse.org/package/show/devel:languages:lua/luajit it is not a git repo, but the distribution style pristine tarball + patches.

skriesch commented:
You can also get the updated patch of this PR with:
wget https://github.com/LuaJIT/LuaJIT/pull/631.patch

edsiper commented:
I am not sure about manual patching. 

wondering if using the updated version provided by OpenResty would be a good way to go:

- https://github.com/LuaJIT/LuaJIT/pull/631#issuecomment-1152035641

@tarruda what do you think ?

cryptomilk commented:
I don't understand what your issue with this patchset it. It is working, tested and maintained. I don't understand what your issue with it is.

corsix commented:
There are some changes (such as 119fd1fa or a4c16404) which need to be made to all backends simultaneously. For such changes to be possible, the author of the change needs to be proficient in the assembly language of all backends, and be able to test all backends. The more backends there are, the harder it becomes to make changes of this nature, and I would argue that we're already at breaking point in this regard: the pool of people with the proficiency and the testing setup rounds to approximately just Mike, and adding another backend likely causes that pool to reduce down to nobody, thereby making certain types of change impossible. Even if it didn't cause the pool to reduce down to nobody, it increases the barrier-to-entry for making the pool larger, and that barrier is already way too high. To prevent this problem from worsening, I wouldn't be surprised to see a one-in-one-out approach for backends, for example dropping PPC32 to make space for RV64.

There may be other arguments in favour or arguments against, but at least for me, the above is a major argument against.

cryptomilk commented:
Thanks for the pointers, I will ask my colleagues if they could update the PR to reflect the latest changes.

skriesch commented:
Feedback for improvements is welcome here. But an alternative solution for project Maintainers outside of Luajit should be the usage of luajit packages provided by the Linux distributions in the LinuxONE (OSS) Community Cloud with the 120 days version (https://community.ibm.com/zsystems/l1cc/) or the long-term version for open source projects (https://community.ibm.com/zsystems/form/l1cc-oss-vm-request/).

mcepl commented:
Anybody can use Hercules <http://www.hercules-390.org/> and I know for certain that maintainers of Red Hat, SUSE, and Debian distros have access to real System Z machines.

juru1234 commented:
You can register for System Z VM's here: https://linuxone.cloud.marist.edu/#/register?flag=VM

There is also QEMU emulation available.

iii-i commented:
I did a rebase and ported a4c1640: https://github.com/iii-i/moonjit/releases/tag/s390x-2024.08.19. I will update the PR after I do a bit more testing.
119fd1f is not necessary, since there is no IR -> assembly support yet.

pemensik commented:
There seems to be some missing parts on s390x. At least dnsjit has failed to build on s390x, bug https://bugzilla.redhat.com/show_bug.cgi?id=2323980.

iii-i committed:
Add s390x architecture support to DynASM

s390x (IBM Z) is an architecture of server computers produced by IBM.
It is supported by a number of open source code generators, such as
GCC, LLVM, OpenJDK, eBPF, QEMU, Valgrind and Cranelift. One of the
missing pieces in the ecosystem support is LuaJIT.

The s390x support for LuaJIT was initially developed by @ketank-new,
@mundaym and @niravthakkar. It found its way into moonjit and luajit2
forks, as well as Fedora distro (as a patch). There were also smaller
contributions by @preetikhorjuvenkar, @Bisht13, @velemas,
@AlekseiNikiforovIBM, and @iii-i.

This is a cumulative patch of the DynASM changes from this work. It
contains all the contributions squashed together, plus minor stylistic
cleanups.

Signed-off-by: Ilya Leoshkevich <iii@linux.ibm.com>

iii-i committed:
Add s390x architecture support

This is a cumulative patch that adds the s390x LuaJIT implementation
by @ketank-new, @mundaym and @niravthakkar and others. It contains all
their contributions squashed together, plus minor stylistic cleanups.
It passes all the tests from LuaJIT-test-cleanup, except for
contents.lua, which fails on x86_64 as well.

Signed-off-by: Ilya Leoshkevich <iii@linux.ibm.com>

aarnez commented:
- Rebased.
- Added support for the new bit manipulation bytecodes required by commit a2ce8114f107464473070427e69e84f4923bd08a.
- Ported two s390x-specific fixes from luajit2.

@sharkz, I believe this should fix the issue with "undefined opcode BC_BNOT". Please let me know if this works for you.
